blog-migration.js

```diff
@@ -0,0 +1,113 @@
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import { readFile, writeFile, mkdir } from 'fs/promises';
+import { glob } from 'glob';
+import { fetchMarkdown } from './fetch-markdown.js';
+import { getMdast, getTableMap, getNodesByType } from './utils/mdast-utils.js';
+import { mdast2docx } from '@adobe/helix-md2docx';
+
+const PROJECT = 'bacom-blog';
+const SITE = 'https://main--business-website--adobe.hlx.page';
+const INDEX = '/blog/query-index.json?limit=3000';
+const USE_CACHE = true;
+const FORCE_SAVE = false;
+
+const MD_DIR = 'md';
+const OUTPUT_DIR = 'output';
+const REPORT_DIR = 'reports';
+
+// TODO: Optimize Save DOCX
+async function saveDocx(mdast, outputFolder, outputFile, force = false) {
+    const output = `${outputFolder}/${outputFile}`;
+    await mkdir(outputFolder, { recursive: true });
+
+    if (force) {
+        const buffer = await mdast2docx(mdast);
+        await writeFile(output, buffer);
+        return;
+    }
+
+    try {
+        await readFile(output);
+        console.log(`Skipping ${output} as docx already exists.`);
+        return;
+    } catch (err) {
+        const buffer = await mdast2docx(mdast);
+        await writeFile(output, buffer);
+    }
+}
+
+export async function main(index, cached, output, force) {
+    const reportDir = `${REPORT_DIR}/${PROJECT}`;
+    const mdDir = `${MD_DIR}/${PROJECT}`;
+    const outputDir = `${output}/${PROJECT}`;
+    const totals = { success: 0, failed: 0 };
+    const failures = [];
+
+    const indexUrl = `${SITE}${index}`;
+    const mdReport = await fetchMarkdown(PROJECT, SITE, indexUrl, cached);
+
+    console.log('totals', mdReport.totals);
+    console.log('failures', mdReport.failures);
+    const mdReportFile = `${reportDir}/markdown.json`;
+    await mkdir(reportDir, { recursive: true });
+    await writeFile(mdReportFile, JSON.stringify(mdReport, null, 2));
+    console.log(`Report written to ${mdReportFile}`);
+
+    const entries = glob.sync(`./${MD_DIR}/${PROJECT}/**/*.md`);
+    for (let i = 0; i < entries.length; i++) {
+        const entry = entries[i];
+        try {
+            const markdown = await readFile(entry, 'utf8');
+            const mdast = await getMdast(markdown);
+
+            // TODO: Migration Part 1 - Pull Quote
+
+            // TODO: Migration Part 2 - Embed
+
+            // TODO: Migration Part 3 - Images
+
+            const outputFolder = `${entry.replace(mdDir, outputDir).split('/').slice(0, -1).join('/')}`;
+            const outputFile = `${entry.split('/').slice(-1)[0].split('.').slice(0, -1).join('.')}.docx`;
```

Comment:
maybe could add a comment on what the expected string is from these methods?

Reply:
Good call, this is a little convoluted, plus we don't need the templating on `outputFolder`. 
I'll break this down for clarity.